### PR TITLE
feat: add FenceLayer costmap plugin to dynamically restrict operational zone

### PR DESCRIPTION
**Pull request by Rudyy75**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #6348 |
| Primary OS tested on | (Ubuntu) |
| Robotic platform tested on | None (Tested via Unit Tests) |
| Does this PR contain AI generated software? | Yes |
| Was this PR description generated by AI software? | No

---

## Description of contribution in a few bullet points

- Added a new `GeofenceLayer` costmap plugin to dynamically restrict the global costmap bounds without needing a static map or map server.
- Created a new ROS 2 service `nav2_msgs/srv/SetFence.srv` to allow updating the operational zone polygon at runtime with full TF2 frame transformation support.
- Used a containment rasterization model: the exterior is marked lethal, the interior is rasterized to free space via scanline fill, and the perimeter edges are raytraced lethal with `nav2_util::raytraceLine`.
- Added a `resize_to_fence` parameter to optionally expand the master costmap to fit the geofence bounds (padded safely by the circumscribed radius).
- Added support for initial fence configuration via the `fence_polygon` parameter, with runtime updates handled via the `SetFence` service. Dynamic parameter callbacks validate and toggle `enabled` and `resize_to_fence`.

## Description of documentation updates required from your changes

- A new plugin documentation page needs to be added to `docs.nav2.org` under Configuration -> Costmap Plugins detailing the `GeofenceLayer`.
- The docs should cover the `enabled`, `resize_to_fence`, and `fence_polygon` parameters, plus an example of calling the `SetFence` service dynamically at runtime.

## Description of how this change was tested

- Wrote a Google Test suite (`geofence_layer_test.cpp`) with 20 unit tests (passes locally in a ROS 2 Rolling Docker environment).
- Tested bounding box updates, TF transforms, and costmap modifications with a rectangular fence.
- Tested the `SetFence` service callback directly to ensure runtime updates and clearing work correctly.
---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in docs.nav2.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists
- [ ] Should this be backported to current distributions? If so, tag with `backport-*`.

### CODE REVIEW

File: nav2_costmap_2d/include/nav2_costmap_2d/fence_layer.hpp [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+#include "nav2_costmap_2d/layered_costmap.hpp"
+#include "nav2_msgs/srv/set_fence.hpp"
+#include "geometry_msgs/msg/point32.hpp"
+
+namespace nav2_costmap_2d
+{
+
+/**
+ * @class FenceLayer
+ * @brief A costmap layer that defines an operational zone (fence).
```
Comment (SteveMacenski) [inline nav2_costmap_2d/include/nav2_costmap_2d/fence_layer.hpp]: This is not what the ticket was asking for, persay. It was to set the size of the costmap based on a service call for the costmap's size when we are not working with a static layer to set the costmap size based on an actual SLAM map or geo tile. 

With that said, it is not a bad idea to have another layer to set a geofenced zone within an existing map of a certain size (and resize the map if required) based on operational zones within the map!

File: nav2_costmap_2d/include/nav2_costmap_2d/fence_layer.hpp [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+ * Cells outside the fence polygon are marked as LETHAL_OBSTACLE,
+ * preventing the planner from routing outside the zone.
+ * The fence can be set via a ROS parameter or updated dynamically
+ * at run-time through a SetFence service.
+ */
+class FenceLayer : public Layer
```
Comment (SteveMacenski) [inline nav2_costmap_2d/include/nav2_costmap_2d/fence_layer.hpp]: Then I Think this should be renamed as a geofence layer

File: nav2_costmap_2d/plugins/fence_layer.cpp [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+namespace nav2_costmap_2d
+{
+
+FenceLayer::FenceLayer()
+{
+}
+
+FenceLayer::~FenceLayer()
+{
+}
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/fence_layer.cpp]: You can use default constructor/destructor then :-)

File: nav2_costmap_2d/plugins/fence_layer.cpp [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+      fence_changed_ = true;
+      RCLCPP_INFO(
+        logger_,
+        "FenceLayer: Loaded fence polygon from parameters with %zu vertices",
+        fence_polygon_.size());
+    } else {
+      RCLCPP_ERROR(
+        logger_,
+        "FenceLayer: Invalid fence_polygon parameter. "
+        "Must be a flat list of doubles [x1,y1,x2,y2,...] with >= 3 vertices.");
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/fence_layer.cpp]: Should this throw or something?

File: nav2_costmap_2d/plugins/fence_layer.cpp [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+void
+FenceLayer::activate()
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  // Add callback for dynamic parameters
+  post_set_params_handler_ = node->add_post_set_parameters_callback(
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/fence_layer.cpp]: What about the validation callback? See other layers for examples. 

Also add the fence itself as a dynamic parameter :-)

File: nav2_costmap_2d/plugins/fence_layer.cpp [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+    response->success = false;
+    response->message = "Fence polygon must have at least 3 vertices";
+    RCLCPP_WARN(
+      logger_,
+      "FenceLayer: Rejected fence with %zu vertices (need >= 3)",
+      polygon.points.size());
+    return;
+  }
+
+  // TODO(future): If request->fence.header.frame_id differs from global_frame_,
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/fence_layer.cpp]: TODO?

File: nav2_costmap_2d/plugins/fence_layer.cpp [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+    master->mapToWorld(
+      master->getSizeInCellsX(), master->getSizeInCellsY(), wx1, wy1);
+
+    *min_x = std::min(*min_x, wx0);
+    *min_y = std::min(*min_y, wy0);
+    *max_x = std::max(*max_x, wx1);
+    *max_y = std::max(*max_y, wy1);
+
+    fence_changed_ = false;
+  }
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/fence_layer.cpp]: Check on the static layer how we might handle if we need to resize the master grid itself because the geofence is larger than the costmap's correct size. I'd add a runtime parameter option for whether this layer should be allowed to resize the costmap or not (for example, we probably wouldn't for rolling costmaps, but probably would for static ones)

Comment (SteveMacenski) [reply nav2_costmap_2d/plugins/fence_layer.cpp]: Any handling we need to do w.r.t. rolling costmaps?

File: nav2_costmap_2d/plugins/fence_layer.cpp [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+    for (int i = min_i; i < max_i; ++i) {
+      master_grid.mapToWorld(
+        static_cast<unsigned int>(i), static_cast<unsigned int>(j), wx, wy);
+      if (!isInsideFence(wx, wy)) {
+        master_grid.setCost(
+          static_cast<unsigned int>(i), static_cast<unsigned int>(j),
+          LETHAL_OBSTACLE);
+      }
+    }
+  }
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/fence_layer.cpp]: This is very brute force and inefficient. We should instead raytrace the geofence boundary and set costs with some thickness of cells (biased towards the outside of the polygon, not 1-cell inside, 1-cell on the line, 1-cell outside) so we only visit and set cells that actually need to change

File: nav2_costmap_2d/plugins/fence_layer.cpp [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+
+  setCurrent(true);
+}
+
+bool
+FenceLayer::isInsideFence(double wx, double wy) const
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/fence_layer.cpp]: Then this can probably be replaced :-)

File: nav2_costmap_2d/plugins/fence_layer.cpp [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+      (wx < (xj - xi) * (wy - yi) / (yj - yi) + xi))
+    {
+      inside = !inside;
+    }
+  }
+  return inside;
+}
+
+bool
+FenceLayer::parsePolygonParameter(
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/fence_layer.cpp]: Do we not have similar we can use for handling footprints?

**Comment by SteveMacenski** [comment]
Per usual - this needs full test coverage and a docs.nav2.org documentation page for configuraiton and navigation plugins guide (but we can handle that after refining the PR a bit more)

**Comment by Rudyy75** [comment]
Hey @SteveMacenski,

Thanks for all the detailed feedback. I went through the comments and updated the PR based on them:

- Renamed to GeofenceLayer: All files and classes have been renamed as requested.
- Performance boost: I got rid of the brute force ray-casting in `updateCosts`. Now, whenever the polygon updates, it just pre-rasterizes everything into a boolean mask. updateCosts just does a quick flat array lookup now, so it should be fast.
- Dynamic resizing: Added the `resize_to_fence` parameter to automatically expand the master costmap to fit the geofence (with padding for the circumscribed radius). It also auto disables itself if it detects a rolling costmap.
- Better parameter handling: Swapped out the manual parser for `makeFootprintFromString`. I also hooked up `on_set_parameters_callback` so we have proper validation, and made `fence_polygon` a dynamic parameter so it can be updated on the fly.
- TF Support: Added the frame transformation logic inside the service callback so it handles different frames correctly now.
- Testing: Wrote a much larger test suite (11 tests now) to make sure all these edge cases, param validation, and resets are thoroughly covered.

@LonelyGuy-SE1, I incorporated a lot of your suggestions as well, thanks for the ideas and guidance :)

Let me know if I've missed anything from the earlier feedback, happy to iterate on it.

File: nav2_costmap_2d/include/nav2_costmap_2d/geofence_layer.hpp [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+ * The geofence can be set via a ROS parameter at startup or updated dynamically
+ * at run-time through the SetFence service. Optionally, the costmap can be
+ * resized to fit the geofence polygon (useful for static costmaps; disabled
+ * for rolling costmaps).
+ */
+class GeofenceLayer : public CostmapLayer
+{
+public:
+  GeofenceLayer() = default;
+  ~GeofenceLayer() = default;
```
Comment (SteveMacenski) [inline nav2_costmap_2d/include/nav2_costmap_2d/geofence_layer.hpp:52]: doxygen :-)

File: nav2_msgs/srv/SetFence.srv [hunk 12]
```diff
@@ -0,0 +1,8 @@
+# Set or update the operational fence polygon for the FenceLayer costmap plugin.
+# The fence defines the operational zone; cells outside are marked as lethal.
```
Comment (SteveMacenski) [inline nav2_msgs/srv/SetFence.srv]:
```suggestion
```

File: nav2_costmap_2d/costmap_plugins.xml [hunk 13]
```diff
@@ -27,6 +27,9 @@
     <class type="nav2_costmap_2d::PluginContainerLayer" base_class_type="nav2_costmap_2d::Layer">
       <description>Plugin to group different layers within the same costmap</description>
     </class>
+    <class type="nav2_costmap_2d::GeofenceLayer" base_class_type="nav2_costmap_2d::Layer">
+      <description>Defines an operational geofence zone; cells outside the polygon are marked lethal. Supports TF-aware polygon updates via the SetFence service and optional costmap resizing.</description>
```
Comment (SteveMacenski) [inline nav2_costmap_2d/costmap_plugins.xml:31]: All cells, or just a border of cells?

Remove "Supports TF-aware..."

File: nav2_costmap_2d/plugins/geofence_layer.cpp [hunk 14]
```diff
@@ -0,0 +1,509 @@
...
+
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"GeofenceLayer: Failed to lock node"};
+  }
+
+  enabled_ = node->declare_or_get_parameter(name_ + ".enabled", true);
+  resize_to_fence_ = node->declare_or_get_parameter(name_ + ".resize_to_fence", true);
+
+  // For rolling costmaps, resizing is not appropriate — warn and override.
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/geofence_layer.cpp:50]: Review all comments and such, these should be human written and readable. This is clearly AI :_)

File: nav2_costmap_2d/plugins/geofence_layer.cpp [hunk 15]
```diff
@@ -0,0 +1,509 @@
...
+}
+
+void
+GeofenceLayer::matchSize()
+{
+  if (!layered_costmap_) {
+    return;
+  }
+  CostmapLayer::matchSize();
+}
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/geofence_layer.cpp:164]: Please review the static layer for how it implements these functions. In this case we should process https://github.com/ros-navigation/navigation2/blob/main/nav2_costmap_2d/plugins/static_layer.cpp#L273-L278

File: nav2_costmap_2d/plugins/geofence_layer.cpp [hunk 16]
```diff
@@ -0,0 +1,509 @@
...
+
+  // Advertise the SetFence service
+  set_fence_service_ = node->create_service<nav2_msgs::srv::SetFence>(
+    name_ + "/set_fence",
+    std::bind(
+      &GeofenceLayer::setFenceCallback, this,
+      std::placeholders::_1, std::placeholders::_2, std::placeholders::_3));
+
+  matchSize();
+  setCurrent(true);
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/geofence_layer.cpp:104]: I don't think we do this in the static layer

File: nav2_costmap_2d/plugins/geofence_layer.cpp [hunk 17]
```diff
@@ -0,0 +1,509 @@
...
+  }
+  on_set_params_handler_.reset();
+  if (post_set_params_handler_ && node) {
+    node->remove_post_set_parameters_callback(post_set_params_handler_.get());
+  }
+  post_set_params_handler_.reset();
+}
+
+void
+GeofenceLayer::reset()
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/geofence_layer.cpp:136]:
```

void
StaticLayer::reset()
{
  has_updated_data_ = true;
  setCurrent(false);
}

```

Why is this set to true and reset maps?

File: nav2_costmap_2d/plugins/geofence_layer.cpp [hunk 18]
```diff
@@ -0,0 +1,509 @@
...
+  // Transform to global_frame_ if needed
+  geometry_msgs::msg::PolygonStamped transformed;
+  const std::string & src_frame = request->fence.header.frame_id;
+  if (!src_frame.empty() && src_frame != global_frame_) {
+    if (!tf_) {
+      response->success = false;
+      response->message = "TF buffer not available; cannot transform polygon";
+      RCLCPP_ERROR(logger_, "GeofenceLayer: %s", response->message.c_str());
+      return;
+    }
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/geofence_layer.cpp:206]: Remove this, tf is always available.

File: nav2_costmap_2d/plugins/geofence_layer.cpp [hunk 19]
```diff
@@ -0,0 +1,509 @@
...
+    try {
+      auto tf_stamped = tf_->lookupTransform(
+        global_frame_, src_frame, tf2::TimePointZero,
+        tf2::durationFromSec(0.5));
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/geofence_layer.cpp:210]: Use a proper TF tolerance

File: nav2_costmap_2d/plugins/geofence_layer.cpp [hunk 20]
```diff
@@ -0,0 +1,509 @@
...
+        }
+        {
+          std::lock_guard<std::mutex> lock(polygon_buffer_mutex_);
+          polygon_buffer_ = poly;
+        }
+        polygon_updated_.store(true);
+        RCLCPP_INFO(
+          logger_,
+          "GeofenceLayer: Updated geofence polygon from parameter with %zu vertices",
+          pts.size());
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/geofence_layer.cpp:503]: This is all duplicate; break into a function used in both places

File: nav2_costmap_2d/plugins/geofence_layer.cpp [hunk 21]
```diff
@@ -0,0 +1,509 @@
...
+  unsigned int span = master_grid.getSizeInCellsX();
+
+  for (int j = min_j; j < max_j; ++j) {
+    for (int i = min_i; i < max_i; ++i) {
+      // Bounds-check against the pre-rasterized mask dimensions
+      if (i < 0 || i >= static_cast<int>(mask_size_x_) ||
+        j < 0 || j >= static_cast<int>(mask_size_y_))
+      {
+        continue;
+      }
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/geofence_layer.cpp:406]: Huh?

File: nav2_costmap_2d/plugins/geofence_layer.cpp [hunk 22]
```diff
@@ -0,0 +1,509 @@
...
+      // Flat array lookup — O(1) per cell, no floating-point math
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/geofence_layer.cpp:407]: I do not understand why this is a comment; is this PR entirely AI generated?

File: nav2_costmap_2d/plugins/geofence_layer.cpp [hunk 23]
```diff
@@ -0,0 +1,509 @@
...
+  for (int i = 0, j = n - 1; i < n; j = i++) {
+    if (((polygon_y_[i] > wy) != (polygon_y_[j] > wy)) &&
+      (wx < (polygon_x_[j] - polygon_x_[i]) * (wy - polygon_y_[i]) /
+      (polygon_y_[j] - polygon_y_[i]) + polygon_x_[i]))
+    {
+      inside = !inside;
+    }
+  }
+  return inside;
+}
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/geofence_layer.cpp:352]: Again, you shouldn't need to do anything this inefficient and brute force, your comments said you resolved this. Raytrace the line segments using the costmap raytracing libraries & come up with a method to inflate outside of the zone so that the border of the geofence has some thickness (which is outside of the polygon, not inside)

File: nav2_costmap_2d/include/nav2_costmap_2d/geofence_layer.hpp [hunk 24]
```diff
@@ -0,0 +1,199 @@
...
+   */
+  void updateParametersCallback(const std::vector<rclcpp::Parameter> & parameters);
+
+  // Polygon state (consumed in updateBounds, protected by polygon_buffer_mutex_)
+  /// Thread-safe buffer holding the next polygon to process
+  std::shared_ptr<const geometry_msgs::msg::PolygonStamped> polygon_buffer_;
+  mutable std::mutex polygon_buffer_mutex_;
+  /// Set to true by the service callback when a new polygon is available
+  std::atomic<bool> polygon_updated_{false};
+
```
Comment (SteveMacenski) [inline nav2_costmap_2d/include/nav2_costmap_2d/geofence_layer.hpp:166]: This seems overengineered; see how other layers do this.

File: nav2_costmap_2d/include/nav2_costmap_2d/geofence_layer.hpp [hunk 25]
```diff
@@ -0,0 +1,199 @@
...
+  /// Polygon vertices (in global_frame_)
+  std::vector<double> polygon_x_;
+  std::vector<double> polygon_y_;
+  /// Pre-rasterized mask: outside_fence_[j * mask_size_x_ + i] == true → LETHAL
+  std::vector<bool> outside_fence_;
+  unsigned int mask_size_x_{0};
+  unsigned int mask_size_y_{0};
+  double mask_origin_x_{0.0};
+  double mask_origin_y_{0.0};
+  double mask_resolution_{0.0};
```
Comment (SteveMacenski) [inline nav2_costmap_2d/include/nav2_costmap_2d/geofence_layer.hpp:177]: I'm not really sure that most of this is necessary.

**Review by SteveMacenski — COMMENTED** [review]
There seems to be issues that would be best if you reviewed the static layer first and made sure you did the right things in the right places (and justify any discrepancies for how this layer is unique). It seems a bit AI  generated without actually appreciating other things that already exist in the codebase and the requested changes. Please make sure that you're implementing this to the standard and consistency you would yourself approve

**Comment by Rudyy75** [comment]
Hey @SteveMacenski, thanks for the detailed review. You're right, I leaned too heavily on AI tooling here and it shows. I've gone back and studied the StaticLayer implementation properly.

I'm reworking the layer to follow the existing CostmapLayer conventions, including the matchSize/reset and parameter handling patterns. For the fence itself, I'll use the existing costmap raytracing utilities to trace the polygon edges, then add a configurable thickness outward from the fence rather than filling the entire outside region.

For the outward thickness, I'm considering expanding the traced boundary into the outside region based on the local polygon edge geometry, with the thickness controlled by a `border_thickness` parameter. Is that direction ok, or is there an existing costmap utility/pattern you'd prefer me to use for the outward expansion?

**Comment by SteveMacenski** [comment]
There's not currently a utility for that, so that's something that you'll need to toy around with :-) 

Sounds good!

**Comment by Rudyy75** [comment]
Hey Steve, sorry for the delay. Been sick for the last week but I'm back at my keyboard now. I plan to have these fixes pushed by saturday :)

**Comment by Rudyy75** [comment]
Hi @SteveMacenski,

Following up on this, offsetting arbitrary polygon edges cleanly around acute/concave corners gets really messy without pulling in an external library, so I stuck with the containment model: exterior cells are filled lethal, the interior is scanline-rasterized to free space, and the perimeter is raytraced with `nav2_util::raytraceLine`. Much simpler, and it means the planner just won't route outside the fence.

Also brought the rest in line with `CostmapLayer` conventions (`matchSize`, `resetMaps`, lifecycle callbacks) and added unit tests. 

CI is green, ready for another look whenever :)

File: nav2_costmap_2d/include/nav2_costmap_2d/fence_layer.hpp (hunk 1 above)
Comment (Rudyy75) [reply nav2_costmap_2d/include/nav2_costmap_2d/fence_layer.hpp]: Understood, I took that second path and focused this PR entirely on the `GeofenceLayer` for operational keep-in zones (including the optional map resize to fit the fence).

File: nav2_costmap_2d/plugins/fence_layer.cpp (hunk 7 above)
Comment (Rudyy75) [reply nav2_costmap_2d/plugins/fence_layer.cpp]: Two things added here:
1) Added a dynamic `resize_to_fence` parameter (defaults to true). If enabled, it resizes the master grid via `layered_costmap_->resizeMap()` to enclose the fence bounds (padded by the circumscribed radius), and restores the original dimensions if the fence is cleared.
2) For rolling costmaps: explicitly rejected them in `onInitialize()` with a runtime error, since a fixed operational geofence doesn't make sense on a moving local window.

File: nav2_costmap_2d/plugins/fence_layer.cpp (hunk 8 above)
Comment (Rudyy75) [reply nav2_costmap_2d/plugins/fence_layer.cpp]: Reworked this completely:

- Moved all the polygon math out of `updateCosts`. It now runs only once when the fence changes and caches to an internal grid, so `updateCosts` just does a fast `updateWithMax`.
- Replaced the per-cell check with a scanline fill for the interior and `nav2_util::raytraceLine` along the edges. Outward buffering got messy around sharp corners without external libraries, so this keeps it fast and robust.

File: nav2_costmap_2d/include/nav2_costmap_2d/geofence_layer.hpp (hunk 11 above)
Comment (Rudyy75) [reply nav2_costmap_2d/include/nav2_costmap_2d/geofence_layer.hpp]: Added!

File: nav2_costmap_2d/costmap_plugins.xml (hunk 13 above)
Comment (Rudyy75) [reply nav2_costmap_2d/costmap_plugins.xml]: All exterior cells are marked lethal

File: nav2_costmap_2d/plugins/geofence_layer.cpp (hunk 14 above)
Comment (Rudyy75) [reply nav2_costmap_2d/plugins/geofence_layer.cpp]: Cleaned up!
